File: 904-fruit-into-baskets/904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        
        unordered_map<int, int> basket;
        
        int leftFruitIdx = 0;
        
        int maxCount = 0;
        
        int count = 0;
        
        for(int i = 0; i < fruits.size(); i++) {
            int curFruit = fruits[i];
            basket[curFruit]++;
            count++;
            
            while(basket.size() > 2) {
                int leftFruit = fruits[leftFruitIdx];
                
                basket[leftFruit]--;
                count--;
                
                if(basket[leftFruit] == 0) {
                    basket.erase(leftFruit);
                }
                
                leftFruitIdx++;
            }
            
            maxCount = max(maxCount, count);
            
        }
        
        return maxCount;
        
    }
};

File: 904-fruit-into-baskets/904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        hashmap = {}
        
        winStart = 0
        
        maxFruit = 0
        
        for winEnd in range(len(fruits)):
            if fruits[winEnd] not in hashmap:
                hashmap[fruits[winEnd]] = 0
                
            hashmap[fruits[winEnd]]+= 1
            
            while len(hashmap) > 2:
                hashmap[fruits[winStart]]-= 1
                if hashmap[fruits[winStart]] == 0:
                    del hashmap[fruits[winStart]]
                winStart+= 1
            
            maxFruit = max(maxFruit, winEnd - winStart + 1)
        
        return maxFruit

# Time Complexity: O(n)
# Space Complexity: O(1) since hashmap will always at most be size k+1 no matter how large the input is